### PR TITLE
Cache refactor

### PR DESCRIPTION
**Pull request by tweekmonster**
Continuing conversation from #70 

@blueyed 

> For entries without a docstring the preview window stays open (which makes sense to avoid flickering, but is not cleared). When selecting an entry with docstring it flickers though (cannot really say if that was the case before though).

I can't say either.  I don't use the preview pane for completions, but I assumed that the buffer not clearing is an issue out of my control.  I guess it might be possible to set the `info` key to a single space to see if that counts as something that can replace the preview contents.

> Then I was not getting any completions anymore (after suspend/hibernate - could that have an influence?!):
> It seems to recover after the "Broken pipe", but there's a delay and it keeps like that.
I suggest creating a PR from your branch to speak on/about the code?!

I'm not sure what's up with that.  What OS are you using?  Could you perhaps reproduce it by doing a bunch of completions, then intentionally suspending?

### CODE REVIEW

**Comment by blueyed** [comment]
> What OS are you using?

Arch Linux.

> Could you perhaps reproduce it by doing a bunch of completions, then intentionally suspending?

Well, it was just a guess - I cannot really say if it failed before and have not checked if it's related to changes here.  Doing a suspend/resume cycle for debugging sounds annoying though to be honest.. :D

**Comment by tweekmonster** [comment]
Hah, fair enough.  We can just note it as a curiosity for now if it's not consistent.

Pure speculation: If suspension is causing this, it might be because of a signal that's not being caught and the process is killed because the server didn't do something in response, or the suspension somehow caused the server's stdin to close and it stopped without an error.  The latter doesn't make sense since neovim's host process was still a live, though.

**Comment by blueyed** [comment]
master causes the same issue, but with different log entries:
```
2016-09-29 22:00:47,568 INFO     (deoplete.logging) --- Deoplete Log Start ---
2016-09-29 22:00:47,568 INFO     (deoplete.logging) NVIM v0.1.6-89-g2daf54e, Python 3.5.2, neovim client 0.1.9
2016-09-29 22:00:51,173 DEBUG    (deoplete.jedi.cache) Input: "user=admin_us"
2016-09-29 22:00:51,174 DEBUG    (deoplete.jedi) Key: ('493a6c7022a908e78abcf351631a62a9', ('conftest', 'admin_user_and_renter_of_another_org'), 'vars'), Refresh: True, Wait: True
2016-09-29 22:00:51,175 DEBUG    (deoplete.jedi.Thread-1) Got work
2016-09-29 22:00:51,175 DEBUG    (deoplete.jedi.Thread-1) Worker error
Traceback (most recent call last):
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/worker.py", line 67, in run
    self.out_queue.put(self.completion_work(*work), timeout=0.5)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/worker.py", line 34, in completion_work
    filename)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/server.py", line 531, in completions
    stream_write(self._server.stdin, args)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/server.py", line 100, in stream_write
    buffer.write(header + data)
BrokenPipeError: [Errno 32] Broken pipe
2016-09-29 22:00:53,179 DEBUG    (deoplete.jedi) Refreshing boilerplate
2016-09-29 22:00:53,180 INFO     (deoplete.jedi) gather_candidates t = 2013.306024ms, µ = 0, σ = 0)
2016-09-29 22:00:53,185 DEBUG    (deoplete.jedi.Thread-1) Got work
2016-09-29 22:00:53,185 DEBUG    (deoplete.jedi.Thread-1) Worker error
Traceback (most recent call last):
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/worker.py", line 67, in run
    self.out_queue.put(self.completion_work(*work), timeout=0.5)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/worker.py", line 34, in completion_work
    filename)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/server.py", line 531, in completions
    stream_write(self._server.stdin, args)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/server.py", line 101, in stream_write
    pipe.flush()
BrokenPipeError: [Errno 32] Broken pipe
```

Then again:
```
2016-09-29 22:01:01,839 DEBUG    (deoplete.jedi.cache) Input: "user=admin_us"
2016-09-29 22:01:01,839 DEBUG    (deoplete.jedi) Key: ('493a6c7022a908e78abcf351631a62a9', ('conftest', 'admin_user_and_renter_of_another_org'), 'vars'), Refresh: True, Wait: True
2016-09-29 22:01:01,840 DEBUG    (deoplete.jedi.Thread-1) Got work
2016-09-29 22:01:01,840 DEBUG    (deoplete.jedi.Thread-1) Worker error
Traceback (most recent call last):
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/worker.py", line 67, in run
    self.out_queue.put(self.completion_work(*work), timeout=0.5)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/worker.py", line 34, in completion_work
    filename)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/server.py", line 531, in completions
    stream_write(self._server.stdin, args)
  File "/home/user/.vim/neobundles/deoplete-jedi/rplugin/python3/deoplete/sources/deoplete_jedi/server.py", line 100, in stream_write
    buffer.write(header + data)
BrokenPipeError: [Errno 32] Broken pipe
2016-09-29 22:01:03,846 INFO     (deoplete.jedi) gather_candidates t = 2013.082867ms, µ = 0, σ = 0)
```

I guess I should create a new issue for that?!
It is really bad that the original error does not show up?!
I am using python-neovim 0.1.9-1 still, IIRC there were some fixes regarding stderr in 0.1.10 - might that be related?!

**Comment by tweekmonster** [comment]
Is there anything logged to `/tmp/jedi-server.log` ?

**Comment by tweekmonster** [comment]
> I guess it might be possible to set the info key to a single space to see if that counts as something that can replace the preview contents.

Confirmed that it's possible, but is pointless to do in deoplete-jedi.

screen-shot-2016-09-29-16-09-15

Everything after `SystemExit()` in the screenshot will have a blank `info` key which will cause the preview to not clear when they're highlighted.  This would need to be a fix in either deoplete.nvim for complete consistency, or in Neovim if this is considered a bug.

**Comment by blueyed** [comment]
It seems to be triggered by some file, and then keeps sticking.
If I use `call deoplete#custom#set('jedi', 'debug_enabled', 1)` (which has to be in vimrc / before the source is loaded, right?) it does not happen though! (on master and this PR)

I've pip-installed neovim 0.10 in the virtualenv, and this appeared to fix the errors.

**Comment by tweekmonster** [comment]
@blueyed Where did we settle on this?  I do think it's a bit mysterious that the errors don't occur when logging is enabled, but #65 should at least deal with it hanging.

**Comment by blueyed** [comment]
@tweekmonster 
Since updating the python-client fixed it, I would not dig deeper into that part of the issue: it seems to be related to handlers being used, which is changed when `server --debug` is used (at least in some regard).

This needs to be rebased anyway: maybe let's split it into the refactoring part and then the new option for a new PR?

File: rplugin/python3/deoplete/sources/deoplete_jedi/server.py [hunk 1]
```diff
@@ -420,20 +397,17 @@ def parse_completion(self, comp, cache):
             type_ = comp.type
             desc = comp.description
 
+        if type_ == '':
+            log.debug('zzzzz %r', comp.description)
```
Comment (tweekmonster) [inline rplugin/python3/deoplete/sources/deoplete_jedi/server.py]: Needs to be deleted before merge.

**Comment by tweekmonster** [comment]
> This needs to be rebased anyway: maybe let's split it into the refactoring part and then the new option for a new PR?

Yeah I'll just remove `deoplete#sources#jedi#show_docstring_sig` (and related code) before I merge this later.

**Comment by blueyed** [comment]
@tweekmonster 
Thanks again!

Can you provide the removed code for `deoplete#sources#jedi#show_docstring_sig` in a new PR (or patch/diff), in case you have it laying around - otherwise I could diff it myself probably.
I'd like to give this a try, since I like the preview option in general, but it is a bit annoying for projects without (much) own documentation.

**Comment by tweekmonster** [comment]
Here's a close enough diff.  I did this quickly because I'm out of town for the weekend.  The only real difference is the new option.  The extra lines in the diff is because of my low effort conflict resolution 😜

```diff
diff --git a/rplugin/python3/deoplete/sources/deoplete_jedi.py b/rplugin/python3/deoplete/sources/deoplete_jedi.py
index bbf8e84..2e88b21 100644
--- a/rplugin/python3/deoplete/sources/deoplete_jedi.py
+++ b/rplugin/python3/deoplete/sources/deoplete_jedi.py
@@ -41,6 +41,14 @@ class Source(Base):
             'deoplete#sources#jedi#show_docstring', False)
         self.debug_server = vars.get(
             'deoplete#sources#jedi#debug_server', None)
+        self.show_docstring_sig = vars.get(
+            'deoplete#sources#jedi#show_docstring_sig', True
+        )
+        # Only one worker is really needed since deoplete-jedi has a pretty
+        # aggressive cache.
+        # Two workers may be needed if working with very large source files.
+        self.worker_threads = vars.get(
+            'deoplete#sources#jedi#worker_threads', 1)
         # Only one worker is really needed since deoplete-jedi has a pretty
         # aggressive cache.
         # Two workers may be needed if working with very large source files.
@@ -110,7 +118,8 @@ class Source(Base):
             sig = '{}({})'.format(item['name'], ', '.join(item['params']))
             sig_len = len(sig)
 
-            desc = sig + '\n\n' + desc
+            if self.show_docstring_sig:
+                desc = sig + '\n\n' + desc
 
             if self.statement_length > 0 and sig_len > self.statement_length:
                 params = []
```

**Pull request merged**